File: tests/plan-policy.test.mjs
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import { mkdtemp, rm, symlink, writeFile } from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const privateValue = "fixture-private-plan-value";
const deletionRule = (age) => [
  { action: [{ type: "Delete" }], condition: [{ age }] },
];
const disk = (auto_delete) => [
  { device_name: "data", source: "data", mode: "READ_WRITE", auto_delete },
];

async function fixture(t, resource = {}, extra = {}) {
  const directory = await mkdtemp(path.join(os.tmpdir(), "infra-plan-policy-"));
  t.after(() => rm(directory, { recursive: true, force: true }));
  const file = path.join(directory, "plan.json");
  await writeFile(
    file,
    JSON.stringify({
      format_version: "1.2",
      terraform_version: "1.12.6",
      errored: false,
      resource_changes: [
        {
          mode: "managed",
          type: "future_resource",
          address: privateValue,
          change: { actions: ["update"], before: {}, after: {} },
          ...resource,
        },
      ],
      ...extra,
    }),
  );
  return { directory, file };
}

function run(args, env = {}) {
  return spawnSync("bash", args, {
    cwd: root,
    encoding: "utf8",
    timeout: 10000,
    env: { ...process.env, ...env },
  });
}

const cases = {
  google_storage_bucket: {
    "retention_policy.0.retention_period": [604800, 86400],
    retention_policy: [[{ retention_period: 604800 }], []],
    "retention_policy.0.is_locked": [true, false],
    "versioning.0.enabled": [true, false],
    "soft_delete_policy.0.retention_duration_seconds": [604800, 0],
    lifecycle_rule: [deletionRule(14), deletionRule(1)],
    public_access_prevention: ["enforced", "inherited"],
    uniform_bucket_level_access: [true, false],
  },
  google_secret_manager_secret: { version_destroy_ttl: ["2592000s", "0s"] },
  google_compute_instance_template: { disk: [disk(false), disk(true)] },
  google_compute_instance_group_manager: {
    stateful_disk: [[{ device_name: "data", delete_rule: "NEVER" }], []],
  },
  google_compute_resource_policy: {
    "snapshot_schedule_policy.0.retention_policy.0.max_retention_days": [7, 1],
  },
  google_cloud_scheduler_job: {
    paused: [false, true],
    schedule: ["0 */4 * * *", "0 0 * * *"],
    time_zone: ["UTC", "Europe/Paris"],
  },
  future_resource: {
    deletion_protection: [true, false],
    force_destroy: [false, true],
    deletion_policy: ["PREVENT", "DELETE"],
  },
};

function fieldValue(field, value) {
  return field
    .split(".")
    .reverse()
    .reduce((child, key) => (key === "0" ? [child] : { [key]: child }), value);
}

for (const [type, fields] of Object.entries(cases)) {
  for (const [field, [before, after]] of Object.entries(fields)) {
    test(`plan policy rejects weakened ${type}.${field}`, async (t) => {
      for (const [value, code] of [
        [before, 0],
        [after, 65],
      ]) {
        const { file } = await fixture(t, {
          type,
          change: {
            actions: ["update"],
            before: fieldValue(field, before),
            after: fieldValue(field, value),
          },
        });
        const result = run(["ops/plan-summary.sh", "foundation", file]);
        assert.equal(result.status, code, result.stderr);
        assert.doesNotMatch(
          result.stdout + result.stderr,
          new RegExp(privateValue),
        );
      }
    });
  }
}

for (const status of ["fail", "error", "unknown", "unsupported", null]) {
  test(`plan policy blocks ${status} check assertions`, async (t) => {
    const { file } = await fixture(
      t,
      {},
      {
        checks: [
          {
            status,
            instances: [{ status, problems: [{ message: privateValue }] }],
          },
        ],
      },
    );
    const result = run(["ops/plan-summary.sh", "foundation", file]);
    assert.equal(result.status, 65);
    assert.doesNotMatch(
      result.stdout + result.stderr,
      new RegExp(privateValue),
    );
  });
}

for (const extra of [
  { errored: true },
  { format_version: "2.0" },
  { checks: false },
  { checks: [{ status: "pass", instances: false }] },
  { checks: [{ status: "pass", instances: [{ status: "fail" }] }] },
]) {
  test(`malformed or failed plan cannot pass: ${JSON.stringify(extra)}`, async (t) => {
    const { file } = await fixture(t, {}, extra);
    assert.equal(run(["ops/plan-summary.sh", "foundation", file]).status, 65);
  });
}

test("adding a cleanup rule requires policy review", async (t) => {
  const { file } = await fixture(t, {
    type: "google_storage_bucket",
    change: {
      actions: ["update"],
      before: {},
      after: { lifecycle_rule: deletionRule(1) },
    },
  });
  assert.equal(run(["ops/plan-summary.sh", "bootstrap", file]).status, 65);
});

test("missing updated resource values cannot bypass protection checks", async (t) => {
  const { file } = await fixture(t, {
    change: {
      actions: ["update"],
      before: { deletion_protection: true },
      after: null,
    },
  });
  assert.equal(run(["ops/plan-summary.sh", "foundation", file]).status, 65);
});

for (const after_unknown of [
  true,
  { retention_policy: true },
  { retention_policy: [{ retention_period: true }] },
]) {
  test(`plan policy rejects unknown protected values ${JSON.stringify(after_unknown)}`, async (t) => {
    const value = { retention_policy: [{ retention_period: 604800 }] };
    const { file } = await fixture(t, {
      type: "google_storage_bucket",
      change: {
        actions: ["update"],
        before: value,
        after: value,
        after_unknown,
      },
    });
    assert.equal(run(["ops/plan-summary.sh", "bootstrap", file]).status, 65);
  });
}

test("safe updates and stronger retention pass with unrelated unknown values", async (t) => {
  const { file } = await fixture(
    t,
    {
      type: "google_storage_bucket",
      change: {
        actions: ["update"],
        before: {
          retention_policy: [{ retention_period: 604800 }],
          labels: { version: "old" },
        },
        after: {
          retention_policy: [{ retention_period: 1209600 }],
          labels: { version: "new" },
        },
        after_unknown: { id: true },
      },
    },
    { checks: [{ status: "pass", instances: [{ status: "pass" }] }] },
  );
  assert.equal(run(["ops/plan-summary.sh", "bootstrap", file]).status, 0);
});

test("preserved disks permit unrelated computed fields and disk reordering", async (t) => {
  const boot = { device_name: "boot", auto_delete: true };
  const { file } = await fixture(t, {
    type: "google_compute_instance_template",
    change: {
      actions: ["update"],
      before: { disk: [boot, ...disk(false)] },
      after: { disk: [...disk(false), boot] },
      after_unknown: { disk: [{ disk_size_gb: true }, { source_image: true }] },
    },
  });
  const result = run(["ops/plan-summary.sh", "foundation", file]);
  assert.equal(result.status, 0, result.stderr);
});

for (const actions of [
  ["delete"],
  ["delete", "create"],
  ["create", "delete"],
  ["forget"],
]) {
  test(`future resource ${actions.join("/")} still requires deletion approval`, async (t) => {
    const { file } = await fixture(t, {
      change: {
        actions,
        before: {},
        after: actions.includes("create") ? {} : null,
      },
    });
    assert.equal(run(["ops/plan-summary.sh", "foundation", file]).status, 3);
  });
}

for (const action of ["plan", "assess", "apply", "converge", "drift"]) {
  test(`${action} rejects unsafe settings even with deletion approval`, async (t) => {
    const { directory, file } = await fixture(t, {
      change: {
        actions: ["update"],
        before: { deletion_protection: true },
        after: { deletion_protection: false },
      },
    });
    await symlink(
      path.join(root, "tests/fixtures/fake-tofu.sh"),
      path.join(directory, "tofu"),
    );
    await symlink("/usr/bin/true", path.join(directory, "git"));
    const env = {
      PATH: `${directory}:${process.env.PATH}`,
      FAKE_TOFU_PLAN_JSON: file,
      FAKE_TOFU_PLAN_CODE: "2",
      FAKE_TOFU_FAIL_ACTION: "apply",
      ALLOW_RESOURCE_DELETION: "true",
    };
    const savedPlan = path.join(directory, "saved.tfplan");
    if (action === "apply") await writeFile(savedPlan, "");
    const args = ["ops/tofu-gate.sh", action, "foundation", "fixture-state"];
    if (["plan", "apply"].includes(action)) args.push(savedPlan);
    const result = run(args, env);
    assert.equal(result.status, 65, `${action}: ${result.stderr}`);
  });
}

const releaseSchedules = [
  ["json_keys_rotation[0]", "agora-json-keys-rotation", "10 * * * *"],
  [
    'postgres_backup["authentication"]',
    "agora-postgres-backup-authentication",
    "45 */4 * * *",
  ],
  [
    'postgres_backup["json_keys"]',
    "agora-postgres-backup-json-keys",
    "15 */4 * * *",
  ],
  [
    'postgres_restore["authentication"]',
    "agora-postgres-restore-authentication",
    "45 3 1 * *",
  ],
  [
    'postgres_restore["json_keys"]',
    "agora-postgres-restore-json-keys",
    "15 3 1 * *",
  ],
  ["postgres_backup_monitor[0]", "agora-postgres-backup-monitor", "5 * * * *"],
];

function releaseSchedule([address, name, schedule] = releaseSchedules[0]) {
  const value = {
    name,
    project: "workload-project-prod",
    region: "europe-west1",
    schedule,
    time_zone: "Etc/UTC",
  };
  return {
    address: `google_cloud_scheduler_job.${address}`,
    type: "google_cloud_scheduler_job",
    change: {
      actions: ["update"],
      before: { ...value, paused: false },
      after: { ...value, paused: true },
      after_unknown: { last_attempt_time: true },
    },
  };
}

function candidateVariables() {
  return {
    application_release: { value: { rollout: { phase: "candidate" } } },
    recovery_mode: { value: false },
    workload_project_id: { value: "workload-project-prod" },
    region: { value: "europe-west1" },
    private_fixture: { value: privateValue },
  };
}

for (const schedule of releaseSchedules) {
  test(`release schedule ${schedule[1]} pauses for a candidate and resumes`, async (t) => {
    for (const phase of ["candidate", "active", "rollback"]) {
      const resource = releaseSchedule(schedule);
      const variables = candidateVariables();
      if (phase !== "candidate") {
        variables.application_release.value.rollout.phase = "active";
        resource.change.before.paused = true;
        resource.change.after.paused = false;
      }
      const { file } = await fixture(t, resource, { variables });
      const result = run(["ops/plan-summary.sh", "release", file]);
      assert.equal(result.status, 0, `${phase}: ${result.stderr}`);
      assert.doesNotMatch(
        result.stdout + result.stderr,
        new RegExp(privateValue),
      );
    }
  });
}

const unsafeCandidateSchedules = {
  "another scheduler": (resource) => {
    resource.address = "google_cloud_scheduler_job.unrelated";
  },
  "a child module scheduler": (resource) => {
    resource.address = `module.other.${resource.address}`;
  },
  "a different cloud job": (resource) => {
    resource.change.before.name = resource.change.after.name = "another-job";
  },
  "a renamed job": (resource) => {
    resource.change.after.name = "another-job";
  },
  "a different project": (resource) => {
    resource.change.before.project = resource.change.after.project =
      "other-project-prod";
  },
  "a changed project": (resource) => {
    resource.change.after.project = "other-project-prod";
  },
  "a different region": (resource) => {
    resource.change.before.region = resource.change.after.region =
      "europe-west2";
  },
  "a changed cron": (resource) => {
    resource.change.after.schedule = "0 0 * * *";
  },
  "a changed time zone": (resource) => {
    resource.change.after.time_zone = "Europe/Paris";
  },
  "unknown pause state": (resource) => {
    resource.change.after_unknown.paused = true;
  },
  "unknown identity": (resource) => {
    resource.change.after_unknown.name = true;
  },
  "unknown project": (resource) => {
    resource.change.after_unknown.project = true;
  },
  "unknown region": (resource) => {
    resource.change.after_unknown.region = true;
  },
  "unknown schedule": (resource) => {
    resource.change.after_unknown.schedule = true;
  },
  "unknown time zone": (resource) => {
    resource.change.after_unknown.time_zone = true;
  },
  "unknown resource": (resource) => {
    resource.change.after_unknown = true;
  },
  "missing pause state": (resource) => {
    delete resource.change.after.paused;
  },
  replacement: (resource) => {
    resource.change.actions = ["delete", "create"];
  },
  "weakened deletion protection": (resource) => {
    resource.change.before.deletion_protection = true;
    resource.change.after.deletion_protection = false;
  },
  "active rollout": (_, variables) => {
    variables.application_release.value.rollout.phase = "active";
  },
  "unknown rollout phase": (_, variables) => {
    variables.application_release.value.rollout.phase = "unknown";
  },
  "absent application": (_, variables) => {
    variables.application_release.value = null;
  },
  "missing phase": (_, variables) => {
    delete variables.application_release.value.rollout.phase;
  },
  "recovery state": (_, variables) => {
    variables.recovery_mode.value = true;
  },
  "missing recovery mode": (_, variables) => {
    delete variables.recovery_mode;
  },
  "missing project": (_, variables) => {
    delete variables.workload_project_id;
  },
  "missing region": (_, variables) => {
    delete variables.region;
  },
};

for (const [name, mutate] of Object.entries(unsafeCandidateSchedules)) {
  test(`candidate pause exception rejects ${name}`, async (t) => {
    const resource = releaseSchedule();
    const variables = candidateVariables();
    mutate(resource, variables);
    const { file } = await fixture(t, resource, { variables });
    const result = run(["ops/plan-summary.sh", "release", file], {
      ALLOW_RESOURCE_DELETION: "true",
    });
    assert.equal(result.status, 65, result.stderr);
    assert.doesNotMatch(
      result.stdout + result.stderr,
      new RegExp(privateValue),
    );
  });
}

for (const rootName of ["bootstrap", "foundation"]) {
  test(`candidate variables cannot authorize pauses in ${rootName}`, async (t) => {
    const { file } = await fixture(t, releaseSchedule(), {
      variables: candidateVariables(),
    });
    assert.equal(run(["ops/plan-summary.sh", rootName, file]).status, 65);
  });
}

test("candidate pause still rejects unresolved plan checks", async (t) => {
  const { file } = await fixture(t, releaseSchedule(), {
    variables: candidateVariables(),
    checks: [{ status: "unknown" }],
  });
  assert.equal(run(["ops/plan-summary.sh", "release", file]).status, 65);
});

File: tests/planPolicy_test.go
package tests_test

import (
	_ "embed"
	"fmt"
	"path/filepath"
	"strings"
	"testing"

	"github.com/stretchr/testify/require"
)

//go:embed testdata/protections.yaml
var protections []byte

func fieldValue(path string, value any) object {
	keys := strings.Split(path, ".")
	for index := len(keys) - 1; index >= 0; index-- {
		if keys[index] == "0" {
			value = []any{value}
		} else {
			value = object{keys[index]: value}
		}
	}
	return value.(object)
}

func plan(resourceType string, before, after any) object {
	return object{
		"format_version": "1.2", "terraform_version": "1.12.6", "errored": false,
		"resource_changes": []any{object{
			"mode": "managed", "type": resourceType, "address": privateValue,
			"change": object{"actions": []string{"update"}, "before": before, "after": after},
		}},
	}
}

func resource(value object) object { return value["resource_changes"].([]any)[0].(object) }

func (f *sandbox) summary(t *testing.T, root string, value object, expected int) string {
	t.Helper()
	file := filepath.Join(f.dir, "plan.json")
	writeJSON(t, file, value)
	code, out := f.script(t, "plan-summary", root, file)
	expectCode(t, expected, code, out)
	return out
}

func TestPlanSummary(t *testing.T) {
	t.Parallel()
	for _, testCase := range []struct {
		fixture string
		code    int
		rows    []string
	}{
		{"safe", 0, []string{"create\tgoogle_cloud_run_v2_job\t1\tcurrent", "import\tgoogle_project\t1\tcurrent", "update\tgoogle_cloud_run_v2_service\t1\tcurrent"}},
		{"no-changes", 0, []string{"action\tresource_type\tcount\tgeneration"}},
		{"protected", 3, []string{"delete\tgoogle_project\t1\tdeposed", "forget\tgoogle_secret_manager_secret\t1\tcurrent", "replace\tgoogle_compute_disk\t1\tcurrent"}},
		{"unsupported", 65, []string{"unsupported action combination"}},
	} {
		t.Run(testCase.fixture, func(t *testing.T) {
			t.Parallel()
			f := setup(t)
			code, output := f.script(t, "plan-summary", "foundation", filepath.Join(f.root, "tests/fixtures/plans", testCase.fixture+".json"))
			expectCode(t, testCase.code, code, output)
			for _, row := range testCase.rows {
				require.Contains(t, output, row)
			}
			require.NotContains(t, output, "fixture-")
			require.NotContains(t, output, "no-op\t")
			if testCase.fixture == "no-changes" {
				require.Equal(t, testCase.rows[0]+"\n", output)
			}
		})
	}
}

func TestPlanProtections(t *testing.T) {
	t.Parallel()
	for _, rule := range fixtureYAML[[]struct {
		Type   string
		Fields map[string][]any
	}](t, protections) {
		for path, pair := range rule.Fields {
			t.Run(rule.Type+"/"+path, func(t *testing.T) {
				t.Parallel()
				f := setup(t)
				f.summary(t, "foundation", plan(rule.Type, fieldValue(path, pair[0]), fieldValue(path, pair[0])), 0)
				f.summary(t, "foundation", plan(rule.Type, fieldValue(path, pair[0]), fieldValue(path, pair[1])), 65)
			})
		}
	}
}

func TestPlanRetention(t *testing.T) {
	t.Parallel()
	for _, path := range []string{"retention_policy.0.retention_period", "soft_delete_policy.0.retention_duration_seconds"} {
		for index, testCase := range []struct {
			before, after any
			code          int
		}{
			{604800, 604800, 0},
			{"604800", 604800, 0},
			{604800, "604800", 0},
			{"604800", "1209600", 0},
			{"604800", "86400", 65},
			{604800, "86400", 65},
			{"604800", 86400, 65},
			{0, "0", 0},
			{"0", 0, 0},
			{"604800", nil, 65},
		} {
			t.Run(fmt.Sprintf("%s/Encoding/%d", path, index), func(t *testing.T) {
				t.Parallel()
				setup(t).summary(t, "bootstrap", plan("google_storage_bucket", fieldValue(path, testCase.before), fieldValue(path, testCase.after)), testCase.code)
			})
		}
		for index, value := range []any{true, false, []any{}, object{}, "", "604800s", " 604800", "604800\n", "+604800", "6.048e5", "604800.0", "-1", -1, 0.5, "9007199254740992", 9007199254740992, privateValue} {
			t.Run(fmt.Sprintf("%s/Invalid/%d", path, index), func(t *testing.T) {
				t.Parallel()
				f := setup(t)
				for _, side := range []string{"before", "after"} {
					valuePlan := plan("google_storage_bucket", fieldValue(path, "604800"), fieldValue(path, "604800"))
					nested(resource(valuePlan), "change")[side] = fieldValue(path, value)
					f.summary(t, "bootstrap", valuePlan, 65)
				}
			})
		}
	}
}

func TestPlanChecks(t *testing.T) {
	t.Parallel()
	cases := []struct {
		name, key string
		value     any
	}{
		{"Errored", "errored", true},
		{"UnsupportedFormat", "format_version", "2.0"},
		{"ChecksShape", "checks", false},
		{"InstancesShape", "checks", []any{object{"status": "pass", "instances": false}}},
		{"InstanceFailure", "checks", []any{object{"status": "pass", "instances": []any{object{"status": "fail"}}}}},
	}
	for _, status := range []any{"fail", "error", "unknown", "unsupported", nil} {
		cases = append(cases, struct {
			name, key string
			value     any
		}{fmt.Sprint(status), "checks", []any{object{"status": status, "instances": []any{object{"status": status, "problems": []any{object{"message": privateValue}}}}}}})
	}
	for _, testCase := range cases {
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()
			value := plan("future_resource", object{}, object{})
			value[testCase.key] = testCase.value
			setup(t).summary(t, "foundation", value, 65)
		})
	}
	for _, testCase := range []struct{ address, outer, inner, diagnostic string }{
		{"google_cloud_quotas_quota_preference.cost_cap", "unknown", "unknown", "REGIONAL_QUOTA_SELECTION (unknown, 1 checks)"},
		{privateValue, "pass", "fail", "OTHER_CHECK (fail, 1 checks)"},
		{privateValue, privateValue, "pass", "OTHER_CHECK (invalid, 1 checks)"},
	} {
		t.Run(testCase.diagnostic, func(t *testing.T) {
			t.Parallel()
			value := plan("future_resource", object{}, object{})
			value["checks"] = []any{object{
				"address": object{"to_display": testCase.address, "kind": privateValue}, "status": testCase.outer,
				"instances": []any{object{"status": testCase.inner, "address": object{"instance_key": privateValue, "to_display": privateValue}, "problems": []any{object{"message": privateValue}}}},
			}}
			require.Contains(t, setup(t).summary(t, "foundation", value, 65), testCase.diagnostic)
		})
	}
}

func TestPlanChanges(t *testing.T) {
	t.Parallel()
	for _, testCase := range []struct {
		name   string
		mutate func(object)
		code   int
	}{
		{"MissingAfter", func(p object) { nested(resource(p), "change")["after"] = nil }, 65},
		{"AddedCleanup", func(p object) {
			nested(resource(p), "change", "after")["lifecycle_rule"] = []any{object{"action": []any{object{"type": "Delete"}}, "condition": []any{object{"age": 1}}}}
		}, 65},
		{"StrongerRetention", func(p object) {
			change := nested(resource(p), "change")
			change["after"] = fieldValue("retention_policy.0.retention_period", "1209600")
			nested(change, "before")["labels"] = object{"version": "old"}
			nested(change, "after")["labels"] = object{"version": "new"}
			change["after_unknown"] = object{"id": true}
			p["checks"] = []any{object{"status": "pass", "instances": []any{object{"status": "pass"}}}}
		}, 0},
		{"UnknownResource", func(p object) { nested(resource(p), "change")["after_unknown"] = true }, 65},
		{"UnknownBlock", func(p object) { nested(resource(p), "change")["after_unknown"] = object{"retention_policy": true} }, 65},
		{"UnknownField", func(p object) {
			nested(resource(p), "change")["after_unknown"] = fieldValue("retention_policy.0.retention_period", true)
		}, 65},
	} {
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()
			value := plan("google_storage_bucket", fieldValue("retention_policy.0.retention_period", "604800"), fieldValue("retention_policy.0.retention_period", "604800"))
			testCase.mutate(value)
			setup(t).summary(t, "bootstrap", value, testCase.code)
		})
	}
	for _, actions := range [][]string{{"delete"}, {"delete", "create"}, {"create", "delete"}, {"forget"}} {
		t.Run(strings.Join(actions, "/"), func(t *testing.T) {
			t.Parallel()
			value := plan("future_resource", object{}, object{})
			nested(resource(value), "change")["actions"] = actions
			if len(actions) == 1 {
				nested(resource(value), "change")["after"] = nil
			}
			setup(t).summary(t, "foundation", value, 3)
		})
	}
	for _, testCase := range []struct {
		name          string
		before, after bool
		code          int
	}{
		{"no-op", false, false, 0}, {"lock", false, true, 0}, {"unlock", true, false, 65},
	} {
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()
			bucket := func(locked bool) object {
				return object{
					"force_destroy": false, "public_access_prevention": "enforced", "uniform_bucket_level_access": true,
					"soft_delete_policy": []any{object{"retention_duration_seconds": 0}},
					"lifecycle_rule":     []any{object{"action": []any{object{"type": "Delete"}}, "condition": []any{object{"age": 14}}}},
					"retention_policy":   []any{object{"retention_period": "604800", "is_locked": locked}},
				}
			}
			value := plan("google_storage_bucket", bucket(testCase.before), bucket(testCase.after))
			if testCase.name == "no-op" {
				nested(resource(value), "change")["actions"] = []string{"no-op"}
			}
			require.NotContains(t, setup(t).summary(t, "bootstrap", value, testCase.code), "604800")
		})
	}
	t.Run("ReorderedDisks", func(t *testing.T) {
		t.Parallel()
		boot := object{"device_name": "boot", "auto_delete": true}
		data := object{"device_name": "data", "source": "data", "mode": "READ_WRITE", "auto_delete": false}
		value := plan("google_compute_instance_template", object{"disk": []any{boot, data}}, object{"disk": []any{data, boot}})
		nested(resource(value), "change")["after_unknown"] = object{"disk": []any{object{"disk_size_gb": true}, object{"source_image": true}}}
		setup(t).summary(t, "foundation", value, 0)
	})
}

func TestPlanServiceExpiration(t *testing.T) {
	t.Parallel()
	for _, testCase := range []struct {
		name, root string
		mutate     func(plan, condition object)
		code       int
	}{
		{"Success", "bootstrap", nil, 0},
		{"Success/AdditionalRestriction", "bootstrap", func(_ object, condition object) { condition["with_state"] = "LIVE" }, 0},
		{"Success/UnchangedRules", "bootstrap", func(p, _ object) {
			change := nested(resource(p), "change")
			change["before"] = change["after"]
		}, 0},
		{"Error/FoundationRoot", "foundation", nil, 65},
		{"Error/ServiceReleaseRoot", "service-release", nil, 65},
		{"Error/OtherAddress", "bootstrap", func(p, _ object) { resource(p)["address"] = "google_storage_bucket.backups" }, 65},
		{"Error/OtherBucket", "bootstrap", func(p, _ object) {
			for _, side := range []string{"before", "after"} {
				nested(resource(p), "change", side)["name"] = "agora-management-test-123456789012-backups"
			}
		}, 65},
		{"Error/OtherProject", "bootstrap", func(p, _ object) {
			for _, side := range []string{"before", "after"} {
				nested(resource(p), "change", side)["project"] = "agora-peer-test"
			}
		}, 65},
		{"Error/MissingRegistration", "bootstrap", func(p, _ object) { delete(p, "variables") }, 65},
		{"Error/RenamedBucket", "bootstrap", func(p, _ object) {
			nested(resource(p), "change", "after")["name"] = "agora-management-test-999999999999-tofu-state"
		}, 65},
		{"Error/ReplacedBucket", "bootstrap", func(p, _ object) {
			nested(resource(p), "change")["actions"] = []string{"delete", "create"}
		}, 65},
		{"Error/UnknownBucket", "bootstrap", func(p, _ object) {
			nested(resource(p), "change")["after_unknown"] = object{"name": true}
		}, 65},
		{"Error/UnknownRule", "bootstrap", func(p, _ object) {
			nested(resource(p), "change")["after_unknown"] = object{"lifecycle_rule": true}
		}, 65},
		{"Error/PrematureExpiry", "bootstrap", func(_ object, condition object) { condition["age"] = 1 }, 65},
		{"Error/MissingPrefix", "bootstrap", func(_ object, condition object) { delete(condition, "matches_prefix") }, 65},
		{"Error/BroadPrefix", "bootstrap", func(_ object, condition object) { condition["matches_prefix"] = []string{"services/", ""} }, 65},
		{"Error/MissingSuffix", "bootstrap", func(_ object, condition object) { delete(condition, "matches_suffix") }, 65},
		{"Error/BroadSuffix", "bootstrap", func(_ object, condition object) { condition["matches_suffix"] = []string{"/plan.tfplan", ".json"} }, 65},
		{"Error/RemovedRule", "bootstrap", func(p, _ object) {
			after := nested(resource(p), "change", "after")
			after["lifecycle_rule"] = after["lifecycle_rule"].([]any)[1:]
		}, 65},
		{"Error/ChangedRule", "bootstrap", func(p, _ object) {
			after := nested(resource(p), "change", "after")
			rule := after["lifecycle_rule"].([]any)[0].(object)
			rule["condition"] = []any{object{"days_since_noncurrent_time": 89, "num_newer_versions": 50}}
		}, 65},
		{"Error/WithdrawCleanup", "bootstrap", func(p, _ object) {
			change := nested(resource(p), "change")
			change["before"], change["after"] = change["after"], change["before"]
		}, 65},
		{"Error/ExtraRule", "bootstrap", func(p, _ object) {
			after := nested(resource(p), "change", "after")
			after["lifecycle_rule"] = append(after["lifecycle_rule"].([]any), fieldValue("action.0.type", "Delete"))
		}, 65},
		{"Error/WeakenedVersioning", "bootstrap", func(p, _ object) {
			nested(resource(p), "change", "after")["versioning"] = []any{object{"enabled": false}}
		}, 65},
		{"Error/WeakenedSoftDelete", "bootstrap", func(p, _ object) {
			nested(resource(p), "change", "after")["soft_delete_policy"] = []any{object{"retention_duration_seconds": 0}}
		}, 65},
	} {
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()
			bucket := func() object {
				return object{
					"name": "agora-management-test-123456789012-tofu-state", "project": "agora-management-test",
					"versioning":         []any{object{"enabled": true}},
					"soft_delete_policy": []any{object{"retention_duration_seconds": 604800}},
					"lifecycle_rule": []any{object{
						"action":    []any{object{"type": "Delete"}},
						"condition": []any{object{"days_since_noncurrent_time": 90, "num_newer_versions": 50}},
					}},
				}
			}
			condition := object{
				"age": 2, "matches_prefix": []string{"services/"},
				"matches_suffix": []string{"/plan.tfplan", "/plan.metadata.json"},
				"with_state":     "ANY", "send_age_if_zero": false,
			}
			before, after := bucket(), bucket()
			after["lifecycle_rule"] = append(after["lifecycle_rule"].([]any), object{
				"action": []any{object{"type": "Delete", "storage_class": ""}}, "condition": []any{condition},
			})
			value := plan("google_storage_bucket", before, after)
			resource(value)["address"] = "google_storage_bucket.state"
			value["variables"] = object{"management_project_id": object{"value": "agora-management-test"}}
			if testCase.mutate != nil {
				testCase.mutate(value, condition)
			}
			setup(t).summary(t, testCase.root, value, testCase.code)
		})
	}
}
